File: app/src/main/java/com/example/myapplication/datastructures/array/codingminutes/linkedlist/LinkedList.java
package com.example.myapplication.datastructures.array.codingminutes.linkedlist;

public class LinkedList<T> {
    private Node head;
    private Node tail;

    public LinkedList() {
        this.head = null;
        this.tail = null;
    }

    public Node getBegin(){
        return head;
    }

    //add element in the front

    public void addFront(T data){
        Node<T> node = new Node<>(data);
        if(head==null){
            head=tail=node;
            node.next=null;
        }else{
            node.next = head;
            head = node;
        }
    }

    //add element in the back

    public void addBack(T data){
        Node<T> node = new Node<>(data);
        if(tail==null){
            head=tail=node;
            node.next=null;
        }else{
            tail.next = node;
            tail = node;
        }
    }

    //insert element in the middle
    public void insert(T data, int pos){
        if(pos==0){
            addFront(data);
            return;
        }
        if(pos>=getSize()-1){
            addBack(data);
            return;
        }
        Node<T> node = new Node<>(data);
        Node currentNode = head;
        for (int i=1;i<pos;i++){
            currentNode = currentNode.next;
        }
        node.next = currentNode.next;
        currentNode.next = node;

    }

    //delete element in the front
    public int deleteFront(){
        if(head==null){
            return -1;
        }
        int deleted = (int) head.data;
        head = head.next;
        return deleted;
    }

    //delete element in the back
    public int deleteBack(){
        if(head==null){
            return -1;
        }
        Node currentNode = head;
        Node prevNode = head;
        while (currentNode.next!=null){
            prevNode = currentNode;
            currentNode = currentNode.next;
        }
        int deleted = (int) currentNode.data;
        tail = prevNode;
        prevNode.next = null;
        System.out.println(prevNode);
        return deleted;
    }

    public int deleteAt(int pos){
        if(pos==0){
            return deleteFront();
        }
        if(pos>getSize()){
            return -1;
        }
        Node currentNode = head;
        Node prevNode = head;
        for(int i=1;i<=pos;i++){
            prevNode = currentNode;
            currentNode = currentNode.next;
        }
        int deleted = (int) currentNode.data;
        prevNode.next = currentNode.next;
        return deleted;
    }


  //get the total node
    public int getSize(){
        int count = 0;
        Node currentNode = head;
        while (currentNode!=null){
            count++;
            currentNode = currentNode.next;
        }
        return count;
    }

    //print all elements from the linked list
    public void print(){
//        System.out.print("head"+head);
        System.out.print("[");
        Node currentNode = head;
        while (currentNode!=null){
            System.out.print(currentNode.data+", ");
            currentNode = currentNode.next;
        }
        System.out.print("]");
    }

    @Override
    public String toString() {
        return "LinkedList{" +
                "head=" + head +
                ", tail=" + tail +
                '}';
    }
}

File: app/src/main/java/com/example/myapplication/datastructures/linkedlist/LinkedList.kt
package com.example.myapplication.datastructures.linkedlist

import com.google.gson.Gson


/**
 * 1. Add - front
 * 2. Add -end
 * 3. Add - middle
 * 4. Delete - front
 * 5. Delete - end
 * 6. Delete - middle
 * 7. print elements
 * */
class LinkedList {
    var startNode: Node? = null
    var endNode: Node? = null
    var count: Int = 0

    fun printLinkedListItems() {
        if (startNode == null) {
            return
        }
        var currentNode = startNode
        println("${currentNode?.data}")
        while (currentNode?.nextNode != null) {
            currentNode = currentNode.nextNode
            println("${currentNode?.data}")
        }
        return
    }

    fun addFront(data: Int) {
        val newNode = Node(data)
        if (startNode == null) {
            startNode = newNode
            endNode = newNode
            count++
        } else {
            startNode?.prevNode = newNode
            newNode.nextNode = startNode
            newNode.prevNode = null
            startNode = newNode
            count++
        }
    }

    fun addEnd(data: Int) {
        val newNode = Node(data)
        if (startNode == null) {
            startNode = newNode
            endNode = newNode
            count++
        } else {
            endNode?.nextNode = newNode
            newNode.prevNode = endNode
            newNode.nextNode = null
            endNode = newNode
            count++
        }
    }

    fun addAfterGivenNode(data: Int, givenNodeData: Int) {
        var givenNodeCheck = searchNodeByData(givenNodeData)
        if (givenNodeCheck != null) {
            val newNode = Node(data)
            if (givenNodeCheck == endNode) {
                addEnd(data)
                return
            }
            val temp = givenNodeCheck.nextNode
            givenNodeCheck.nextNode = newNode
            newNode.nextNode = temp
            newNode.prevNode = givenNodeCheck
            count++
        }
        return
    }

    private fun searchNodeByData(data: Int): Node? {
        if (startNode == null) {
            return null
        }
        var currentNode = startNode
        while (currentNode?.nextNode != null || currentNode == endNode) {
            if (currentNode?.data == data) {
                return currentNode
            }
            currentNode = currentNode?.nextNode
        }
        return null
    }

    fun deleteFront(): Node? {
        if (startNode == null) {
            return null
        }
        val tempNode = startNode?.nextNode
        val deletedNode = startNode
        tempNode?.prevNode = null
        startNode = tempNode
        count--
        return deletedNode
    }

    fun deleteEnd(): Node? {
        if (startNode == null) {
            return null
        }
        val endNodePrevNode = endNode?.prevNode
        val deletedNode = endNode
        endNodePrevNode?.nextNode = null
        endNode = endNodePrevNode
        count--
        return deletedNode
    }

    fun deleteNodeBySearch(data:Int){
        val isGivenNodeAvailable = searchNodeByData(data) ?: return
        if(startNode==isGivenNodeAvailable){
            deleteFront()
            return
        }
        if(endNode==isGivenNodeAvailable){
            deleteEnd()
            return
        }
        val previousNode = isGivenNodeAvailable.prevNode
        val nextNode = isGivenNodeAvailable.nextNode
        previousNode?.nextNode = nextNode
        nextNode?.prevNode =  previousNode
    }

    fun getSize() = count
}